clang/lib/Sema/SemaSYCLDeclAttr.cpp

```diff
@@ -3162,3 +3162,88 @@ void SemaSYCL::checkSYCLAddIRAttributesFunctionAttrConflicts(Decl *D) {
       Diag(Attr->getLoc(), diag::warn_sycl_old_and_new_kernel_attributes)
           << Attr;
 }
+
+static bool isFreeFunctionKernel(SemaSYCL &S, FunctionDecl *FD) {
+  /* If it does not have the attribute, return false. */
+  if (!FD->hasAttr<SYCLAddIRAttributesFunctionAttr>())
+    return false;
+  /* If it does, traverse through the pairs and return true, if
+     it is one of the two free function types. */
+  auto *SAIRAttr = FD->getAttr<SYCLAddIRAttributesFunctionAttr>();
+  SmallVector<std::pair<std::string, std::string>, 4> NameValuePairs =
+      SAIRAttr->getFilteredAttributeNameValuePairs(S.getASTContext());
+  for (const auto &NVPair : NameValuePairs) {
+    if (!NVPair.first.compare("sycl-single-task-kernel") ||
+        !NVPair.first.compare("sycl-nd-range-kernel"))
+      return true;
+  }
+  /* If it does not have free function attributes, return false. */
+  return false;
+}
+
+void SemaSYCL::handleSYCLRegisteredKernels(Decl *D, const ParsedAttr &A) {
+  // Check for SYCL device compilation context.
+  if (!getLangOpts().SYCLIsDevice)
+    return;
+
+  unsigned NumArgs = A.getNumArgs();
+  /* When declared, we expect at least one item in the list. */
+  if (NumArgs == 0) {
+    Diag(A.getLoc(), diag::err_registered_kernels_num_of_args);
+    return;
+  }
+
+  /* Traverse through the items in the list. */
+  for (unsigned I = 0; I < NumArgs; I++) {
+    assert(A.isArgExpr(I) && "Expected expression argument");
+    /* Each item in the list must be an initializer list expression. */
+    Expr *ArgExpr = A.getArgAsExpr(I);
+    if (!isa<InitListExpr>(ArgExpr)) {
+      Diag(ArgExpr->getExprLoc(), diag::err_registered_kernels_init_list);
+      return;
+    }
+
+    const auto *ArgListE = cast<InitListExpr>(ArgExpr);
+    unsigned NumInits = ArgListE->getNumInits();
+    /* Each init-list expression must have a pair of values. */
+    if (NumInits != 2) {
+      Diag(ArgExpr->getExprLoc(),
+           diag::err_registered_kernels_init_list_pair_values);
+      return;
+    }
+
+    /* The first value of the pair must be a string. */
+    const Expr *FirstExpr = ArgListE->getInit(0);
+    StringRef CurStr;
+    SourceLocation Loc = FirstExpr->getExprLoc();
+    if (!SemaRef.checkStringLiteralArgumentAttr(A, FirstExpr, CurStr, &Loc))
+      return;
+
+    /* Resolve the FunctionDecl from the second value of the pair. */
+    auto *SecondE = const_cast<Expr *>(ArgListE->getInit(1));
```

Comment:
Do you really need a const cast here? `SecondE` doesn't seem to be changed

Reply:
The cast is needed not for SecondE - though you are right that it is not modified.   The issue is the expressions we pull from SecondE that are passed to routines further below.   For example: ``ResolveSingleFunctionTemplateSpecialization`` does not like a ``const UnresolvedLookupExpr *``.

Reply:
Well that is a bit sad. Do you know why `ArgListE->getInit()` returns `Expr *` with const modifier? Maybe it shouldn't?